### PR TITLE
feat: add configuration & transport structs

### PR DESCRIPTION
**Pull request by aaravm**
This pull request introduces two new structs: `Configuration` for storing API request details and `Transport` for making HTTP requests using the `reqwest` library. It also includes unit tests for the `Transport` struct.

- **New Features**:
    - Introduced a `Configuration` struct to store key details of API requests, including base path, user agent, authentication, and API key information.
    - Added a `Transport` struct that utilizes the `Configuration` struct and a `reqwest::Client` to facilitate making GET, POST, PUT, and DELETE HTTP requests.
- **Tests**:
    - Added unit tests for the `Transport` struct to verify the functionality of HTTP methods using the `mockito` library.

## Summary by Sourcery

This pull request introduces two new structs: `Configuration` for storing API request details and `Transport` for making HTTP requests using the `reqwest` library. It also includes unit tests for the `Transport` struct.

- **New Features**:
    - Introduced a `Configuration` struct to store key details of API requests, including base path, user agent, authentication, and API key information.
    - Added a `Transport` struct that utilizes the `Configuration` struct and a `reqwest::Client` to facilitate making GET, POST, PUT, and DELETE HTTP requests.
- **Tests**:
    - Added unit tests for the `Transport` struct to verify the functionality of HTTP methods using the `mockito` library.

### CODE REVIEW

**Review by pavelnikonorov — APPROVED** [review]
LGTM!

File: lib/src/configuration.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#[derive(Debug, Clone)]
+pub struct Configuration {
```
Comment (uniqueg) [inline lib/src/configuration.rs:2]: Please provide doc comments for this and each other struct in this PR (and in the future), containing some or all of:
- The purpose of the struct
- What each field represents
- Any other notes (special considerations or constraints)

File: lib/src/configuration.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// pub type BasicAuth = (String, Option<String>);
+
+#[derive(Debug, Clone)]
+pub struct ApiKey {
+    pub prefix: Option<String>,
+    pub key: String,
+}
+
+impl Configuration {
+    pub fn new(
```
Comment (uniqueg) [inline lib/src/configuration.rs:26]: Please add method-level doc comments to this and all other methods, with
- Description of method
- Param descriptions
- Description of return value
- Usage examples / doc (unit) tests

File: lib/src/configuration.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        base_path: String,
```
Comment (uniqueg) [inline lib/src/configuration.rs:27]: Could perhaps use https://crates.io/crates/url?

**Review by uniqueg — CHANGES_REQUESTED** [review]
I think documentation and unit tests are missing.

File: lib/src/configuration.rs [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#[derive(Debug, Clone)]
+pub struct Configuration {
+    pub base_path: String,
+    pub user_agent: Option<String>,
+    pub basic_auth: Option<BasicAuth>,
+    pub oauth_access_token: Option<String>,
+    pub bearer_access_token: Option<String>,
+    pub api_key: Option<ApiKey>,
```
Comment (uniqueg) [inline lib/src/configuration.rs:8]: Do we really want all of these to be public?

File: lib/src/configuration.rs (hunk 2 above)
Comment (vemonet) [reply lib/src/configuration.rs:26]: Use the docstring `///` for every method/function, you can write the docstring in markdown and all codeblocks will be automatically executed when running the tests with `cargo test`. Make sure the whole block is working as standalone

The docstrings will automatically be used to generate the documentation at docs.rs

This way you don't need to write 100s of other tests, you will already have a test for each method/function, and your documentation is sure to be right and up-to-date if the tests pass (this is really useful for future sustainability)

Example docstring with test: https://github.com/vemonet/nanopub-rs/blob/main/lib/src/nanopub.rs#L113

File: lib/src/configuration.rs [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#[derive(Debug, Clone)]
+pub struct ApiKey {
+    pub prefix: Option<String>,
+    pub key: String,
+}
+
+impl Configuration {
+    pub fn new(
+        base_path: String,
+        user_agent: Option<String>,
```
Comment (vemonet) [inline lib/src/configuration.rs:28]: There is no optional args in rust, only nullable one. So constructors with `Option<>` will require to provide all args anyway, e.g. 

```rust
Configuration::new("/path".to_string(), null, null, null)
```

It does not look really nice, but more important in the future if you want to add 1 more optional arguments to your configuration then **all previous use of the constructor will need an update to add one more `null`!**.

Either making your codebase too rigid, or forcing you to introduce breaking changes often

For future sustainability it is recommended to handle constructors with many optional args with the builder pattern: https://rust-unofficial.github.io/patterns/patterns/creational/builder.html

e.g. https://github.com/vemonet/nanopub-rs/blob/main/lib/src/profile.rs#L30

Comment (uniqueg) [reply lib/src/configuration.rs:51]: Interesting, good to know :pray:

File: lib/src/configuration.rs (hunk 3 above)
Comment (aaravm) [reply lib/src/configuration.rs]: I like this idea! Changed

File: lib/src/configuration.rs (hunk 5 above)
Comment (aaravm) [reply lib/src/configuration.rs]: Thanks for the review!
I didn't know about this, and I have now updated this PR to the builder pattern

File: lib/src/configuration.rs (hunk 4 above)
Comment (aaravm) [reply lib/src/configuration.rs]: @uniqueg, I had a talk with @pavelnikonorov, and I think for now, we can keep the parameters public as it isn't a security risk for now, but later on we could use getter and setters

**Comment by aaravm** [comment]
@uniqueg, I think this PR can be merged now. Please mention if you think any other change should be made..

**Review by uniqueg — APPROVED** [review]
LGTM - I think you can go ahead and merge, @aaravm (didn't do it because I'm not sure about the order)

**Comment by aaravm** [comment]
I have merged this PR, as it doesn't really depend on any other PR.

**Pull request merged**